src/palindrom-network-channel.js

```diff
@@ -0,0 +1,365 @@
+import URL from './URLShim';
+import { PalindromError, PalindromConnectionError } from './palindrom-errors';
+/* this package will be empty in the browser bundle,
+and will import https://www.npmjs.com/package/websocket in node */
+import WebSocket from 'websocket';
+import nodeFetch from 'node-fetch';
+
+const CLIENT = 'Client';
+const SERVER = 'Server';
+
+/**
+ * Replaces http and https to ws and wss in a URL and returns it as a string.
+ * @param  {String} remoteUrl HTTP remote address
+ * @return {String}           WS address
+ */
+function toWebSocketURL(remoteUrl) {
+    /* replace 'http' strictly in the beginning of the string,
+    this covers http and https */
+    return remoteUrl.replace(/^http/i, 'ws');
+}
+
+export default class PalindromNetworkChannel {
+    constructor(
+        palindrom,
+        remoteUrl,
+        useWebSocket,
+        onReceive,
+        onSend,
+        onConnectionError,
+        onSocketOpened,
+        onFatalError,
+        onStateChange
+    ) {
+        // TODO(tomalec): to be removed once we will achieve better separation of concerns
+        this.palindrom = palindrom;
+
+        if (typeof window !== 'undefined' && window.location) {
+            this.remoteUrl = new URL(remoteUrl, window.location.href);
+        } else {
+            // in Node, URL is absolute
+            this.remoteUrl = new URL(remoteUrl);
+        }
+
+        onReceive && (this.onReceive = onReceive);
+        onSend && (this.onSend = onSend);
+        onConnectionError && (this.onConnectionError = onConnectionError);
+        onFatalError && (this.onFatalError = onFatalError);
+        onStateChange && (this.onStateChange = onStateChange);
+        onSocketOpened && (this.onSocketOpened = onSocketOpened);
+
+        Object.defineProperty(this, 'useWebSocket', {
```

Comment:
This duplicates the getters and setters from lines 74, 77. I know it was also like that before this PR, but maybe that's an opportunity to fix it?

Reply:
This seems to be not addressed. @tomalec what do we do about it?